unit_tests.py

```diff
@@ -192,11 +196,195 @@ def test_build_multi_fasta(self):
+                      ("s3://s3-csu-001///", "baz")]
+        test_output = ["abc/123/consensus/foo_consensus.fas",
+                       "5/1/consensus/bar_consensus.fas", 
+                       "consensus/baz_consensus.fas"] 
+        fail = False 
+        i = 0
+        for input, output in zip(test_input, test_output):
+            try:
+                self.assertEqual(btb_phylo.extract_s3_key(input[0], input[1]), output)
+            except AssertionError as e:
+                i += 1
+                fail = True
+                print(f"Test failure {i}: ", e)
+        if fail: 
+            print(f"{i} test failures")
 
-if __name__ == "__main__":
-    unittest.main(buffer=True)
+    def test_match_s3_uri(self):
+        # test exceptions
+        with self.assertRaises(btb_phylo.BadS3UriError):
+            btb_phylo.match_s3_uri('s3://s3-csu-003')
+        with self.assertRaises(btb_phylo.BadS3UriError):
+            btb_phylo.match_s3_uri('s3://s3-csu-003abc')
+        with self.assertRaises(btb_phylo.BadS3UriError):
+            btb_phylo.match_s3_uri('s3://s3-csu-03/abc')
+        with self.assertRaises(btb_phylo.BadS3UriError):
+            btb_phylo.match_s3_uri('s3:/s3-csu-003/abc')
+        with self.assertRaises(btb_phylo.BadS3UriError):
+            btb_phylo.match_s3_uri('s4://s3-csu-003/abc')
+        with self.assertRaises(btb_phylo.BadS3UriError):
+            btb_phylo.match_s3_uri('s3://s5-abc-003/abc')
+        with self.assertRaises(btb_phylo.BadS3UriError):
+            btb_phylo.match_s3_uri('s3://s3-csu-abc/abc')
+        with self.assertRaises(btb_phylo.BadS3UriError):
+            btb_phylo.match_s3_uri('s3://s3-csu-1234/abc')
+
+class TestUpdateSummary(unittest.TestCase):
+    def test_extract_submission_no(self):
+        # Test cases
+        test_input = ["AFxx-12-34567-89",
+                      "ATxx-12-34567-89",
+                      "AFx-12-34567-89",
+                      "Ax-12-34567-89",
+                      "AF-12-34567-89",
+                      "AFx12-34567-89",
+                      "HI-12-34567-89",
+                      "12-34567-89-1L",
+                      "12-34567-89-L1",
+                      "A-12-34567-89",
+                      "12-34567-89-1",
+                      "12-34567-89-L",
+                      "12-34567-89",
+                      "AFxx-12-3456-89",
+                      "ATxx-12-3456-89",
+                      "AFx-12-3456-89",
+                      "Ax-12-3456-89",
+                      "AF-12-3456-89",
+                      "AFx12-3456-89",
+                      "HI-12-3456-89",
+                      "12-3456-89-1L",
+                      "12-3456-89-L1",
+                      "A-12-3456-89",
+                      "12-3456-89-1",
+                      "12-3456-89-L",
+                      "12-3456-89",
+                      "12345678",
+                      "ABCDEFGH",
+                      ""]
+        test_output = ["12-34567-89",
```

Comment:
Submission numbers all have an `AF-` prefix in LIMS, but is this what's required for matching with other data sources?

Reply:
Not too sure what you mean, I was just using the submission number (without AF) to choose which samples to include. The snp matrix has the original sample names, i.e. AF-....